### PR TITLE
[CLI config parity] migrate url type to string type

### PR DESCRIPTION
**Pull request by ashmeenkaur**
### Description
This PR includes the following changes:
- removes url type from supported new config types. This was being used by custom end point parameter.
- Refactors StorageClientConfig's CustomEndPoint to string type (from url.Url earlier) for simplification.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - Updated
3. Integration tests - Ran via KOKORO

### CODE REVIEW

File: cfg/validate_test.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+		Compress:        false,
+		MaxFileSizeMb:   1,
+	}
+}
+
+func TestValidateConfig(t *testing.T) {
+	testCases := []struct {
+		name                   string
+		config                 *Config
+		expectedCustomEndpoint string
```
Comment (kislaykishore) [inline cfg/validate_test.go:35]: I think it's good enough to check whether Validate returns an error. This expectedCustomEndpoint doesn't add any value and can be removed.

File: cfg/validate.go [hunk 2]
```diff
@@ -28,10 +29,25 @@ func isValidLogRotateConfig(config *LogRotateLoggingConfig) error {
...
 // ValidateConfig returns a non-nil error if the config is invalid.
 func ValidateConfig(config *Config) error {
-	if err := isValidLogRotateConfig(&config.Logging.LogRotate); err != nil {
+	var err error
+
+	if err = isValidLogRotateConfig(&config.Logging.LogRotate); err != nil {
 		return fmt.Errorf("error parsing log-rotate config: %w", err)
 	}
+
+	if config.GcsConnection.CustomEndpoint, err = isValidUrl(config.GcsConnection.CustomEndpoint); err != nil {
```
Comment (kislaykishore) [inline cfg/validate.go:48]: Let's not change the parsed config in ValidateConfig. It's mandate is to just return an error if the config is invalid.

Comment (ashmeenkaur) [reply cfg/validate.go:48]: Valid concern, but I am not sure what will be the best fix in this case. 

We can (1)move it / (2)duplicate the check to rationalize? 
(1) throwing an error during rationalize doesn't look right
(2) I am not in favor of duplicating

Comment (kislaykishore) [reply cfg/validate.go:48]: I am okay with (2) i.e. duplicate the url.Parse call with a TODO to check whether there is any benefit of parsing the rawURL first and then changing the URL back to raw URL.

File: cfg/validate.go [hunk 3]
```diff
@@ -28,10 +29,25 @@ func isValidLogRotateConfig(config *LogRotateLoggingConfig) error {
 	return nil
 }
 
+func isValidUrl(u string) (string, error) {
```
Comment (kislaykishore) [inline cfg/validate.go:32]: As per Go convention it should be named as isValidURL.

File: cfg/validate_test.go (hunk 1 above)
Comment (ashmeenkaur) [reply cfg/validate_test.go:35]: Added a test case where this is being used. PTAL.

File: cfg/validate.go (hunk 2 above)
Comment (ashmeenkaur) [reply cfg/validate.go]: As discussed offline, rawURL parsing is required (added a unit test for it). We will continue with the current approach of doing the check and assignment in the validation method.

File: cfg/validate.go [hunk 4]
```diff
@@ -28,10 +29,25 @@ func isValidLogRotateConfig(config *LogRotateLoggingConfig) error {
 	return nil
 }
 
+func isValidURL(u string) (string, error) {
```
Comment (gargnitingoogle) [inline cfg/validate.go:32]: The function name is misleading. It sounds like it is only validating the given URI, but it's also trying to parse it and return the successfully parsed url or error. How about 

```suggestion
func decodedURL(u string) (string, error) {
```

or 

```suggestion
func parsedURL(u string) (string, error) {
```

**Review by gargnitingoogle — CHANGES_REQUESTED** [review]
one minor readability comment. Please address.

File: cfg/validate.go (hunk 4 above)
Comment (ashmeenkaur) [reply cfg/validate.go]: @kislaykishore and I had a discussion and we decided to duplicate this in the Rationalise() function keeping the responsibility of validate and rationalise separate. I've taken these suggestions in the rationalise part.

File: cfg/rationalize.go [hunk 5]
```diff
@@ -14,10 +14,28 @@
...
 package cfg
 
+import "net/url"
+
+func decodeURL(u string) (string, error) {
+	decodedUrl, err := url.Parse(u)
+	if err != nil {
+		return "", err
+	}
+	return decodedUrl.String(), nil
```
Comment (kislaykishore) [inline cfg/rationalize.go:24]:
```suggestion
	decodedURL, err := url.Parse(u)
	if err != nil {
		return "", err
	}
	return decodedURL.String(), nil
```

File: cfg/validate.go [hunk 6]
```diff
@@ -28,10 +29,25 @@ func isValidLogRotateConfig(config *LogRotateLoggingConfig) error {
...
 // ValidateConfig returns a non-nil error if the config is invalid.
 func ValidateConfig(config *Config) error {
-	if err := isValidLogRotateConfig(&config.Logging.LogRotate); err != nil {
+	var err error
+
+	if err = isValidLogRotateConfig(&config.Logging.LogRotate); err != nil {
 		return fmt.Errorf("error parsing log-rotate config: %w", err)
 	}
+
+	if err = isValidURL(config.GcsConnection.CustomEndpoint); err != nil {
```
Comment (kislaykishore) [inline cfg/validate.go:48]: Can we reuse the `decodeURL` method here.

File: cmd/legacy_param_mapper.go [hunk 7]
```diff
@@ -35,6 +35,14 @@ func PopulateNewConfigFromLegacyFlagsAndConfig(c cliContext, flags *flagStorage,
 		return nil, fmt.Errorf("PopulateNewConfigFromLegacyFlagsAndConfig: unexpected nil flags or mount config")
 	}
 
+	// resolve custom endpoint url type to string.
+	var customEndPoint string
+	if flags.CustomEndpoint == nil {
+		customEndPoint = ""
+	} else {
+		customEndPoint = flags.CustomEndpoint.String()
+	}
```
Comment (kislaykishore) [inline cmd/legacy_param_mapper.go:44]: This is more succinct.
```suggestion
	var customEndPoint string
	if flags.CustomEndpoint != nil {
		customEndPoint = flags.CustomEndpoint.String()
	} 
```

File: cfg/rationalize.go [hunk 8]
```diff
@@ -14,10 +14,28 @@
 
 package cfg
 
+import "net/url"
+
+func decodeURL(u string) (string, error) {
+	decodedUrl, err := url.Parse(u)
```
Comment (kislaykishore) [inline cfg/rationalize.go:20]: nit: Feel free to ignore - Can we add a TODO to replace `Parse` call with [ParseRequestURI](https://pkg.go.dev/net/url#ParseRequestURI).

**Review by kislaykishore — CHANGES_REQUESTED** [review]

**Review by kislaykishore — APPROVED** [review]

File: cfg/validate_test.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+	}{
+		{
+			name: "Valid Config 1",
+			config: &Config{
+				Logging: LoggingConfig{LogRotate: validLogRotateConfig()},
+				GcsConnection: GcsConnectionConfig{
+					CustomEndpoint: "https://bing.com/search?q=dotnet",
+				},
+			},
+			wantErr: false,
```
Comment (gargnitingoogle) [inline cfg/validate_test.go:45]: nit: We don't need to set the `false` value explicitly as it's default and is obvious.

**Review by gargnitingoogle — COMMENTED** [review]
Added some readability comments/questions.

File: cfg/validate_test.go (hunk 9 above)
Comment (ashmeenkaur) [reply cfg/validate_test.go:45]: I think this is okay for test as it enhances the test readability

File: cfg/validate.go [hunk 10]
```diff
@@ -28,10 +28,22 @@ func isValidLogRotateConfig(config *LogRotateLoggingConfig) error {
 	return nil
 }
 
+func isValidURL(u string) (err error) {
+	_, err = decodeURL(u)
+	return
+}
+
 // ValidateConfig returns a non-nil error if the config is invalid.
 func ValidateConfig(config *Config) error {
```
[the remark this answers is not in the document]
Comment (ashmeenkaur) [reply cfg/validate.go:37]: Will do this in a separate PR to keep the scope of changes limited to PR description.

File: cfg/rationalize_test.go [hunk 11]
```diff
@@ -59,9 +59,64 @@ func TestRationalize(t *testing.T) {
...
+			expectedCustomEndpoint: "",
+			wantErr:                true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actualErr := Rationalize(tc.config)
+
+			assert.Equal(t, tc.config.GcsConnection.CustomEndpoint, tc.expectedCustomEndpoint)
```
[the remark this answers is not in the document]
Comment (ashmeenkaur) [reply cfg/rationalize_test.go:114]: I have separated out success and failure case test based on https://google.github.io/styleguide/go/decisions.html#table-driven-tests

File: cfg/rationalize_test.go [hunk 12]
```diff
@@ -68,55 +68,63 @@ func TestRationalizeEnableEmptyManagedFolders(t *testing.T) {
...
+func TestRationalizeCustomEndpointSuccessful(t *testing.T) {
 	testCases := []struct {
 		name                   string
 		config                 *Config
 		expectedCustomEndpoint string
-		wantErr                bool
 	}{
 		{
-			name: "Valid Config 1",
+			name: "Valid Config where input and expected custom endpoint matches.",
```
Comment (gargnitingoogle) [inline cfg/rationalize_test.go:78]: s/matches/match/

File: cfg/validate_test.go [hunk 13]
```diff
@@ -28,53 +28,61 @@ func validLogRotateConfig() LogRotateLoggingConfig {
...
 	testCases := []struct {
-		name    string
-		config  *Config
-		wantErr bool
+		name   string
+		config *Config
 	}{
 		{
-			name: "Valid Config 1",
+			name: "Valid Config where input and expected custom endpoint matches.",
```
Comment (gargnitingoogle) [inline cfg/validate_test.go:37]: s/matches/match/

**Review by gargnitingoogle — COMMENTED** [review]
comment about minor typo

**Review by gargnitingoogle — APPROVED** [review]
LGTM

**Review by kislaykishore — APPROVED** [review]

**Review by charith87 — APPROVED** [review]

**Pull request merged**